src/value.rs

```diff
@@ -416,90 +695,88 @@ impl Value {
-            match &data.node.inner {
-                ValueInner::Unit => {
-                    if data.n_children_yielded == 0
-                        && !matches!(
-                            data.parent.map(|value| &value.inner),
-                            Some(ValueInner::Left(_)) | Some(ValueInner::Right(_))
-                        )
-                    {
-                        f.write_str("ε")?;
-                    }
-                }
-                ValueInner::Left(..) => {
-                    if data.n_children_yielded == 0 {
-                        f.write_str("0")?;
-                    }
+        // This is a copy of the logic inside `CompactBitsIter` except
+        // that we handle products more explicitly.
+        enum S<'v> {
+            Disp(Value),
+            DispRef(&'v Value),
+            DispUnlessUnit(Value),
+            DispCh(char),
+        }
+
+        let mut stack = Vec::with_capacity(1024);
+        // Next node to visit, and a boolean indicating whether we should
+        // display units explicitly (turned off for sums, since a sum of
+        // a unit is displayed simply as 0 or 1.
+        stack.push(S::DispRef(self));
+
+        while let Some(next) = stack.pop() {
+            let value = match next {
+                S::Disp(ref value) | S::DispUnlessUnit(ref value) => value,
+                S::DispRef(value) => value,
+                S::DispCh(ch) => {
+                    write!(f, "{}", ch)?;
+                    continue;
                 }
-                ValueInner::Right(..) => {
-                    if data.n_children_yielded == 0 {
-                        f.write_str("1")?;
-                    }
+            };
+
+            if value.is_unit() {
+                if !matches!(next, S::DispUnlessUnit(..)) {
+                    f.write_str("ε")?;
                 }
-                ValueInner::Product(..) => match data.n_children_yielded {
-                    0 => f.write_str("(")?,
-                    1 => f.write_str(",")?,
-                    2 => f.write_str(")")?,
-                    _ => unreachable!(),
-                },
+            } else if let Some(l_value) = value.to_left() {
+                f.write_str("0")?;
+                stack.push(S::DispUnlessUnit(l_value));
+            } else if let Some(r_value) = value.to_right() {
+                f.write_str("1")?;
+                stack.push(S::DispUnlessUnit(r_value));
+            } else if let Some((l_value, r_value)) = value.to_product() {
+                stack.push(S::DispCh(')'));
+                stack.push(S::Disp(r_value));
+                stack.push(S::DispCh(','));
+                stack.push(S::Disp(l_value));
+                stack.push(S::DispCh('('));
             }
         }
         Ok(())
     }
 }
-
 /// An iterator over the bits of the padded encoding of a [`Value`].
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
-struct PaddedBitsIter {
+struct CompactBitsIter {
```

Comment:
cbbe9b5a3e7982c5b4c85b28cc37996fbcb0c7b2: This documentation talks about padded encoding, but the iterator handles compact encoding.